File: source/modulo_components/test/python/test_lifecycle_component_communication.py
import pytest
from modulo_components.lifecycle_component import LifecycleComponent
from modulo_utils.exceptions import ModuloError


class Trigger(LifecycleComponent):
    def __init__(self):
        super().__init__("trigger")

    def on_configure_callback(self) -> bool:
        self.add_trigger("test")
        return True

    def on_activate_callback(self) -> bool:
        self.trigger("test")
        return True


@pytest.mark.parametrize("minimal_cartesian_input", [[LifecycleComponent, "/topic"]], indirect=True)
@pytest.mark.parametrize("minimal_cartesian_output", [[LifecycleComponent, "/topic", True]], indirect=True)
def test_input_output(ros_exec, make_lifecycle_change_client, random_pose, minimal_cartesian_output,
                      minimal_cartesian_input):
    input_change_client = make_lifecycle_change_client("minimal_cartesian_input")
    output_change_client = make_lifecycle_change_client("minimal_cartesian_output")
    ros_exec.add_node(input_change_client)
    ros_exec.add_node(output_change_client)
    ros_exec.add_node(minimal_cartesian_input)
    ros_exec.add_node(minimal_cartesian_output)
    input_change_client.configure(ros_exec)
    output_change_client.configure(ros_exec)
    input_change_client.activate(ros_exec)
    output_change_client.activate(ros_exec)
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert minimal_cartesian_input.received_future.done() and minimal_cartesian_input.received_future.result()
    assert random_pose.get_name() == minimal_cartesian_input.input.get_name()
    assert random_pose.dist(minimal_cartesian_input.input) < 1e-3
    with pytest.raises(ModuloError):
        minimal_cartesian_output.publish()


@pytest.mark.parametrize("minimal_cartesian_input", [[LifecycleComponent, "/topic"]], indirect=True)
@pytest.mark.parametrize("minimal_cartesian_output", [[LifecycleComponent, "/topic", False]], indirect=True)
def test_input_output_manual(ros_exec, make_lifecycle_change_client, random_pose, minimal_cartesian_output,
                             minimal_cartesian_input):
    input_change_client = make_lifecycle_change_client("minimal_cartesian_input")
    output_change_client = make_lifecycle_change_client("minimal_cartesian_output")
    ros_exec.add_node(input_change_client)
    ros_exec.add_node(output_change_client)
    ros_exec.add_node(minimal_cartesian_input)
    ros_exec.add_node(minimal_cartesian_output)
    input_change_client.configure(ros_exec)
    output_change_client.configure(ros_exec)
    input_change_client.activate(ros_exec)
    output_change_client.activate(ros_exec)
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert not minimal_cartesian_input.received_future.done()
    minimal_cartesian_output.publish()
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert minimal_cartesian_input.received_future.done() and minimal_cartesian_input.received_future.result()
    assert random_pose.get_name() == minimal_cartesian_input.input.get_name()
    assert random_pose.dist(minimal_cartesian_input.input) < 1e-3


@pytest.mark.parametrize("minimal_cartesian_input", [[LifecycleComponent, "/topic"]], indirect=True)
@pytest.mark.parametrize("minimal_joint_output", [[LifecycleComponent, "/topic", True]], indirect=True)
def test_input_output_invalid_type(ros_exec, minimal_joint_output, minimal_cartesian_input):
    ros_exec.add_node(minimal_cartesian_input)
    ros_exec.add_node(minimal_joint_output)
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert not minimal_cartesian_input.received_future.done()


@pytest.mark.parametrize("minimal_cartesian_input", [[LifecycleComponent, "/topic"]], indirect=True)
def test_input_output_invalid_msg(ros_exec, make_lifecycle_change_client, make_minimal_invalid_encoded_state_publisher,
                              minimal_cartesian_input):
    input_change_client = make_lifecycle_change_client("minimal_cartesian_input")
    invalid_publisher = make_minimal_invalid_encoded_state_publisher("/topic")
    ros_exec.add_node(input_change_client)
    ros_exec.add_node(minimal_cartesian_input)
    ros_exec.add_node(invalid_publisher)
    input_change_client.configure(ros_exec)
    input_change_client.activate(ros_exec)
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert not minimal_cartesian_input.received_future.result()


def test_trigger(ros_exec, make_lifecycle_change_client, make_predicates_listener):
    trigger = Trigger()
    listener = make_predicates_listener("/trigger", ["test"])
    client = make_lifecycle_change_client("trigger")
    ros_exec.add_node(trigger)
    ros_exec.add_node(listener)
    ros_exec.add_node(client)
    client.configure(ros_exec)
    ros_exec.spin_until_future_complete(listener.predicates_future, timeout_sec=0.5)
    assert not listener.predicates_future.done()
    assert not listener.predicate_values["test"]
    client.activate(ros_exec)
    ros_exec.spin_until_future_complete(listener.predicates_future, timeout_sec=0.5)
    assert listener.predicates_future.done()
    assert listener.predicate_values["test"]

File: source/modulo_components/test/cpp/test_lifecycle_component_communication.cpp
#include <gtest/gtest.h>

#include <modulo_utils/testutils/PredicatesListener.hpp>

#include "modulo_components/LifecycleComponent.hpp"
#include "test_modulo_components/communication_components.hpp"

using namespace modulo_components;

class LifecycleTrigger : public LifecycleComponent {
public:
  explicit LifecycleTrigger(const rclcpp::NodeOptions& node_options) : LifecycleComponent(node_options, "trigger") {}

  bool on_configure_callback() final {
    this->add_trigger("test");
    return true;
  }

  bool on_activate_callback() final {
    this->trigger("test");
    return true;
  }
};

class LifecycleComponentCommunicationTest : public ::testing::Test {
protected:
  void SetUp() override {
    rclcpp::init(0, nullptr);
    exec_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
  }

  void TearDown() override {
    rclcpp::shutdown();
  }

  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> exec_;
};

TEST_F(LifecycleComponentCommunicationTest, InputOutput) {
  auto cartesian_state = state_representation::CartesianState::Random("test");
  auto input_node = std::make_shared<MinimalCartesianInput<LifecycleComponent>>(rclcpp::NodeOptions(), "/topic");
  auto output_node = std::make_shared<MinimalCartesianOutput<LifecycleComponent>>(
      rclcpp::NodeOptions(), "/topic", cartesian_state, true);
  add_configure_activate(this->exec_, input_node);
  add_configure_activate(this->exec_, output_node);
  auto return_code = this->exec_->spin_until_future_complete(input_node->received_future, 500ms);
  ASSERT_EQ(return_code, rclcpp::FutureReturnCode::SUCCESS);
  EXPECT_EQ(cartesian_state.get_name(), input_node->input->get_name());
  EXPECT_TRUE(cartesian_state.data().isApprox(input_node->input->data()));
  EXPECT_THROW(output_node->publish(), modulo_components::exceptions::ComponentException);
}

TEST_F(LifecycleComponentCommunicationTest, InputOutputManual) {
  auto cartesian_state = state_representation::CartesianState::Random("test");
  auto input_node = std::make_shared<MinimalCartesianInput<LifecycleComponent>>(rclcpp::NodeOptions(), "/topic");
  auto output_node = std::make_shared<MinimalCartesianOutput<LifecycleComponent>>(
      rclcpp::NodeOptions(), "/topic", cartesian_state, false);
  add_configure_activate(this->exec_, input_node);
  add_configure_activate(this->exec_, output_node);
  auto return_code = this->exec_->spin_until_future_complete(input_node->received_future, 500ms);
  ASSERT_EQ(return_code, rclcpp::FutureReturnCode::TIMEOUT);
  output_node->publish();
  return_code = this->exec_->template spin_until_future_complete(input_node->received_future, 500ms);
  ASSERT_EQ(return_code, rclcpp::FutureReturnCode::SUCCESS);
  EXPECT_EQ(cartesian_state.get_name(), input_node->input->get_name());
  EXPECT_TRUE(cartesian_state.data().isApprox(input_node->input->data()));
}

TEST_F(LifecycleComponentCommunicationTest, Trigger) {
  auto trigger = std::make_shared<LifecycleTrigger>(rclcpp::NodeOptions());
  auto listener =
      std::make_shared<modulo_utils::testutils::PredicatesListener>("trigger", std::vector<std::string>{"test"});
  this->exec_->add_node(trigger->get_node_base_interface());
  trigger->configure();
  this->exec_->add_node(listener);
  auto result_code = this->exec_->spin_until_future_complete(listener->get_predicate_future(), 500ms);
  ASSERT_EQ(result_code, rclcpp::FutureReturnCode::TIMEOUT);
  EXPECT_FALSE(listener->get_predicate_values().at("test"));
  trigger->activate();
  result_code = this->exec_->spin_until_future_complete(listener->get_predicate_future(), 500ms);
  ASSERT_EQ(result_code, rclcpp::FutureReturnCode::SUCCESS);
  EXPECT_TRUE(listener->get_predicate_values().at("test"));
}
